numpy/array_api/linalg.py

```diff
@@ -0,0 +1,371 @@
+
+def solve(x1: Array, x2: Array, /) -> Array:
+    """
+    Array API compatible wrapper for :py:func:`np.solve <numpy.solve>`.
+
+    See its docstring for more information.
+    """
+    # Note: the restriction to floating-point dtypes only is different from
+    # np.linalg.solve.
+    if x1.dtype not in _floating_dtypes or x2.dtype not in _floating_dtypes:
+        raise TypeError('Only floating-point dtypes are allowed in solve')
+
+    return Array._new(np.linalg.solve(x1._array, x2._array))
+
+def svd(x: Array, /, *, full_matrices: bool = True) -> SVDResult:
+    """
+    Array API compatible wrapper for :py:func:`np.svd <numpy.svd>`.
+
+    See its docstring for more information.
+    """
+    # Note: the restriction to floating-point dtypes only is different from
+    # np.linalg.svd.
+    if x.dtype not in _floating_dtypes:
+        raise TypeError('Only floating-point dtypes are allowed in svd')
+
+    # Note: the return type here is a namedtuple, which is different from
+    # np.svd, which only returns a tuple.
+    return SVDResult(*map(Array._new, np.linalg.svd(x._array, full_matrices=full_matrices)))
+
+# Note: svdvals is not in NumPy (but it is in SciPy). It is equivalent to
+# np.linalg.svd(compute_uv=False).
+def svdvals(x: Array, /) -> Union[Array, Tuple[Array, ...]]:
+    return Array._new(np.linalg.svd(x._array, compute_uv=False))
+
+# Note: tensordot is the numpy top-level namespace but not in np.linalg
+
+# Note: axes must be a tuple, unlike np.tensordot where it can be an array or array-like.
+def tensordot(x1: Array, x2: Array, /, *, axes: Union[int, Tuple[Sequence[int], Sequence[int]]] = 2) -> Array:
+    # Note: the restriction to numeric dtypes only is different from
+    # np.tensordot.
+    if x1.dtype not in _numeric_dtypes or x2.dtype not in _numeric_dtypes:
+        raise TypeError('Only numeric dtypes are allowed in tensordot')
+
+    return Array._new(np.tensordot(x1._array, x2._array, axes=axes))
+
+# Note: trace is the numpy top-level namespace, not np.linalg
+def trace(x: Array, /, *, offset: int = 0) -> Array:
+    """
+    Array API compatible wrapper for :py:func:`np.trace <numpy.trace>`.
+
+    See its docstring for more information.
+    """
+    # Note: trace always operates on the last two axes, whereas np.trace
+    # operates on the first two axes by default
+    return Array._new(np.asarray(np.trace(x._array, offset=offset, axis1=-2, axis2=-1)))
+
+# Note: vecdot is not in NumPy
+def vecdot(x1: Array, x2: Array, /, *, axis: Optional[int] = None) -> Array:
+    if axis is None:
+        axis = -1
+    return tensordot(x1, x2, axes=((axis,), (axis,)))
+
+
+# Note: the name here is different from norm(). The array API norm is split
+# into matrix_norm and vector_norm().
+def vector_norm(x: Array, /, *, axis: Optional[Union[int, Tuple[int, int]]] = None, keepdims: bool = False, ord: Optional[Union[int, float, Literal[np.inf, -np.inf]]] = 2) -> Array:
```

Comment:
This followed PyTorch: [`tensordot`](https://pytorch.org/docs/stable/generated/torch.tensordot.html) uses `dims` and [`vector_norm`](https://pytorch.org/docs/stable/generated/torch.linalg.vector_norm.html) (and [`matrix_norm`](https://pytorch.org/docs/stable/generated/torch.linalg.matrix_norm.html)) uses `dim`.

Admittedly, this is, at first blush, somewhat inconsistent. Typically, we've tried to use the singular for when a single axis can be specified (e.g., see the [manipulation functions](https://github.com/data-apis/array-api/blob/f267ebd3505bc13d9f49569021669b6742bcd046/spec/API_specification/manipulation_functions.md), which have some overlap in terms of `axis` usage).

We did not do that for `tensordot` as the `axes` keyword does not behave the same as the other APIs. When an `int`, axes specifies the number of axes over which to contract, not the axis index.

Reply:
IMO matrix_norm should use axes and not axis. Although I didn't realize the manipulation functions do this too, so maybe it's not as inconsistent as it seems.

Actually, I'm not even sure why matrix_norm has an axis argument at all. It's the only function here that lets you specify the axes that matrices are stacked on. This is done for some vector APIs (`tensordot`, `cross`, and `vector_norm`), but none of the matrix APIs.

Reply:
The `axis` kwarg for `matrix_norm` follows PyTorch which is the only library which had implemented this API prior to standardization. We could remove from the specification, making the default behavior the last two axes. PyTorch could still support the `axis` kwarg without being non-compliant.

Reply:
It also comes from `np.linalg.norm`. It was probably implemented there for consistency for the vector case, but IMO it makes less sense now that the function is split since no other matrix function does this (unless we want to add `axis` to the other matrix functions).

Reply:
`axis` keyword to `matrix_norm` was removed (xref https://github.com/data-apis/array-api/issues/318).

`tensordot` was reviewed and kept as "int or tuple of 2 sequences".